### ISSUE TITLE
Model translation warns about failed assignments and fails with -d=dumpSimCode

### ISSUE THREAD

**Issue by rfranke** [OpenModelica/OpenModelica#10519]
### Description

The model `ThermofluidStream.Examples.SimpleAirCycle`, as well as 11 more models of that library, raise warnings of the form:
```
[1] 10:26:00 Translation Warning
SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_2.outletA.state.X[1], splitterT1_2.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source_outlet_Medium_ThermodynamicState1.X

[2] 10:26:00 Translation Warning
SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_1.outletA.state.X[1], splitterT1_1.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState7.X

[3] 10:26:00 Translation Warning
SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_2.outletA.state.X[1], splitterT1_2.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source_outlet_Medium_ThermodynamicState391.X

[4] 10:26:00 Translation Warning
SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_1.outletA.state.X[1], splitterT1_1.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState477.X
```

The translation fails when `-d=dumpSimCode` is specified.
```
Translation Error
[/.../OpenModelica/OMCompiler/Compiler/SimCode/SimCodeUtil.mo: 798:5-798:146]: Internal error function createSimCode failed [Transformation from optimised DAE to simulation code structure failed]
```

Equations like
```
{splitterT1_2.outletA.state.X[1], splitterT1_2.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source_outlet_Medium_ThermodynamicState1.X
```
might better be formulated as
```
splitterT1_2.outletA.state.X = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source_outlet_Medium_ThermodynamicState1.X
```
to fit into `SimCodeUtil.makeSES_SIMPLE_ASSIGN` and to improve the generated code (no need to create a temporary array `{X[1],X[2]}` for the lhs when directly assigning to `X`).

### Steps to Reproduce

Simulate the model to get the warnings. Alternatively the smaller `ThermofluidStream.Processes.Tests.Nozzle`
Specify `-d=dumpSimCode` to make the translation fail.

### Expected Behavior

The models should translate without warning and dumpSimCode should work.

### Version and OS

The current master branch.

**Comment by casella** [comment]
But it doesn't work with cpp runtime 😞

**Comment by kabdelhak** [comment]
Can't reproduce the mentioned behavior that the model runs fine. For me it just fails, also with C-runtime.

The problematic part was a weird record assignment. It failed because record attributes that are arrays were not thought of apparently, i fixed it in the PR, now they are parsed correctly, but i still get an error:
```
ThermofluidStream.Examples.SimpleAirCycle.c:1641:6: note: previous definition of ‘ThermofluidStream_Examples_SimpleAirCycle_eqFunction_471’ was here
 1641 | void ThermofluidStream_Examples_SimpleAirCycle_eqFunction_471(DATA *data, threadData_t *threadData)
```
Multiple redefinitions of functions, i will see what is up here.

**Comment by rfranke** [comment]
Strange. For me the C simulation runs with version 1.0.0 of ThermofluidStream (as can be installed from OMEdit). Using the current master branch:
```
loadModel(ThermofluidStream); getErrorString();
simulate(ThermofluidStream.Examples.SimpleAirCycle); getErrorString();
```
gives
```
true
""
record SimulationResult
    resultFile = "/home/rfranke/work/ThermofluidStream.Examples.SimpleAirCycle_res.mat",
    simulationOptions = "startTime = 0.0, stopTime = 100.0, numberOfIntervals = 1000, tolerance = 1e-06, method = 'dassl', fileNamePrefix = 'ThermofluidStream.Examples.SimpleAirCycle', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = ''",
    messages = "LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_SUCCESS       | info    | The simulation finished successfully.
",
    timeFrontend = 1.174497638,
    timeBackend = 0.457996175,
    timeSimCode = 0.07743319,
    timeTemplates = 0.13148054,
    timeCompile = 6.534537113,
    timeSimulation = 1.096674818,
    timeTotal = 9.472781424000001
end SimulationResult;
"Warning: SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_2.outletA.state.X[1], splitterT1_2.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source_outlet_Medium_ThermodynamicState1.X
Warning: SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_1.outletA.state.X[1], splitterT1_1.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState7.X
Warning: SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_2.outletA.state.X[1], splitterT1_2.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source_outlet_Medium_ThermodynamicState391.X
Warning: SimCodeUtil.makeSES_SIMPLE_ASSIGN failed for: {splitterT1_1.outletA.state.X[1], splitterT1_1.outletA.state.X[2]} = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState477.X
"
```

**Comment by kabdelhak** [comment]
> Trying out your PR,
> 
>     * the warnings disappear
> 
>     * dumpSimCode still crashes
> 
>     * C++ fails as before
> 
>     * C now also fails as reported by you

dumpSimCode does not crash for me, very weird...

**Comment by kabdelhak** [comment]
Oh nevermind, i see now what i did wrong. I think my PR is part of the solution, but there is more to it.

**Comment by kabdelhak** [comment]
I fixed the broken simcode dump in #10581. It could not handle printing a function pointer being passed to a function, i added the missing cases and also made a backup case of <unknown> to prevent missing cases from failing entirely in the future.

Back on the original problem now.

**Comment by kabdelhak** [comment]
Attempts to fix this problem are in PR #10604 

With this PR, the model generates code, but i might have made errors in the code changes. I just added lines of code so that it would first of all not fail during code generation.

Running it with the current version of the PR yields following error 4 times:
```
OMCppThermofluidStream.Examples.SimpleAirCycle.h:255:131: error: expected parameter name, found \"2\"
  255 |     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_X_(2) _pointerToRealVars[61]
      |    
```
My assumption is, that this error is just the result of the compiler trying to parse an array variable while still having a subscript or dimension indicator at the end.

After removing these with dirty changes just to see what happens and if it can actually compile, the following error occurs approx 100 times:
```
OMCppThermofluidStream.Examples.SimpleAirCycleAlgloop666.cpp:74:12: error: ‘class SimpleAirCycle’ has no member named ‘_junctionT1_1_P_junctionN_P_r_in_1’; did you mean ‘_junctionT1_1_P_junctionN_P_r_in_’?
   74 |   _system->_junctionT1_1_P_junctionN_P_r_in_1 = ((_system->_junctionT1_1_P_outlet_P_state_P_p_ + _system->_junctionT1_1_P_junctionN_P_r_mix_) - _system->_junctionT1_1_P_junctionN_P_p_1);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Which seems like there is an extra `1` at the end of each name. This might be a result of my dirty fix, but could lead to the core of the problem.

My understanding so far of this problem is that we have records containing arrays, which are tried to be parsed as scalar because it does not expect arrays here (i had to add code to SimCode to make it work). Sine this should be a common structure there is probably something special about this record (individual record variable bindings, special constructor?). As a result the naming of the variables seems to be off. What would be the expected code at the mentioned places?

**Comment by rfranke** [comment]
Thank you for your effort so far @kabdelhak!
There is some confusion between arrays and array elements. The model header file `OMCppThermofluidStream.Examples.SimpleAirCycle.h` defines the first array element as array of dim 1 and the second array element as scalar (define into overall vector of variables):
```C++
    StatArrayDim1<double, 1, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_X_;
    #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_X_(2) _pointerToRealVars[61]
```
Only one can be right: either array of dim 2 or expanded element with with `2` at the end instead of `(2)`.

dumpSimCode says:
```
algVars (367)
----------------------
...
index:5: $TMP_11[1] (no alias)  protected  hideResult  initial: 	arrCref:$TMP_11 index:(34) [1]
index:6: $TMP_120[1] (no alias)  protected  hideResult  initial: 	arrCref:$TMP_120 index:(35) [1]
index:7: $TMP_132[1] (no alias)  protected  hideResult  initial: 	arrCref:$TMP_132 index:(36) [1]
...
index:32: $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState474.X[1] (no alias)  protected  hideResult  initial: 	arrCref:$TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState474.X index:(61) []
index:33: $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState474.X[2] (no alias)  protected  hideResult  initial: 	no arrCref index:(62) []
...
```
`ThermodynamicState474.X[1]` is "arrCref", whereas `ThermodynamicState474.X[2]` is "no arrCref". Is this correct?
Also the $TMP arrays of dim 1 don't appear efficient, but this is another issue and they might be needed.

A similar confusion, but the other way around happens in Algloop666. The model header file declares an array of dim 2:
```C++
    StatArrayDim1<double, 2, true> _junctionT1_1_P_junctionN_P_r_in_;
```
The first element should be accessed with `_junctionT1_1_P_junctionN_P_r_in_(1)`.
dumpSimCode declares them similarly:
```
index:268: junctionT1_1.junctionN.r_in[1] (no alias)  protected  initial: 	arrCref:junctionT1_1.junctionN.r_in index:(297) [2]
index:269: junctionT1_1.junctionN.r_in[2] (no alias)  protected  initial: 	no arrCref index:(298) [2]
```

Should this be treated in the SimCode or the Templates?

**Comment by rfranke** [comment]
@kabdelhak it looks that you have solved the second problem with Algloop666 already.
If I fix the remaining first problem of bad declarations of four TMP arrays manually, everything compiles and simulates!
`_junctionT1_1_P_junctionN_P_r_in_` still serves as good example for generated SimCode (see above).
```diff
--- a/OMCppThermofluidStream.Examples.SimpleAirCycle.generated.h
+++ b/OMCppThermofluidStream.Examples.SimpleAirCycle.h
@@ -251,20 +251,16 @@ protected:
     StatArrayDim1<double, 1, true> __omcQ_24TMP_5F594_;
     StatArrayDim1<double, 1, true> __omcQ_24TMP_5F72_;
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_T_ _pointerToRealVars[59]
-    StatArrayDim1<double, 1, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_X_;
-    #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_X_(2) _pointerToRealVars[61]
+    StatArrayDim1<double, 2, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_X_;
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState474_P_p_ _pointerToRealVars[62]
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState6_P_T_ _pointerToRealVars[63]
-    StatArrayDim1<double, 1, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState6_P_X_;
-    #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState6_P_X_(2) _pointerToRealVars[65]
+    StatArrayDim1<double, 2, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState6_P_X_;
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState6_P_p_ _pointerToRealVars[66]
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState1_P_T_ _pointerToRealVars[67]
-    StatArrayDim1<double, 1, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState1_P_X_;
-    #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState1_P_X_(2) _pointerToRealVars[69]
+    StatArrayDim1<double, 2, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState1_P_X_;
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState1_P_p_ _pointerToRealVars[70]
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState389_P_T_ _pointerToRealVars[71]
-    StatArrayDim1<double, 1, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState389_P_X_;
-    #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState389_P_X_(2) _pointerToRealVars[73]
+    StatArrayDim1<double, 2, true> __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState389_P_X_;
     #define __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource_5Foutlet_5FMedium_5FThermodynamicState389_P_p_ _pointerToRealVars[74]
     #define __omcQ_24cse1_ _pointerToRealVars[75]
     #define __omcQ_24cse10_ _pointerToRealVars[76]
```

**Comment by rfranke** [comment]
PR #10630 adds the missing array dimensions. This seems to have been the root cause as the issue disappears without further changes.

**Comment by rfranke** [comment]
PR #10632 changes the above assignment of individual array elements to a combined
```C++
/*
equation index: 474
type: SIMPLE_ASSIGN
splitterT1_1.outletA.state.X = $TMP_ThermofluidStream_Examples_SimpleAirCycle_source1_outlet_Medium_ThermodynamicState471.X
*/
void SimpleAirCycle::evaluate_474()
{
  //Array assign
  _splitterT1_1_P_outletA_P_state_P_X_ = __omcQ_24TMP_5FThermofluidStream_5FExamples_5FSimpleAirCycle_5Fsource1_5Foutlet_5FMedium_5FThermodynamicState471_P_X_;
}
```

**Comment by rfranke** [comment]
... the generated assignments don't mach in all cases without expansion. Expanding everything matches them brute force. But that's another issue. This one is solved with PR #10630.

**Comment by rfranke** [comment]
Thank's to the work by @kabdelhak on dumpSimCode, we could better analyse the problem and then solve it.
